src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -111,50 +143,46 @@ public bool TryResolveReferences(MetadataReferenceInfo metadataReferenceInfo, [N
 
         public bool TryGetAssemblyInfo(MetadataReferenceInfo metadataReferenceInfo, [NotNullWhen(true)] out AssemblyInfo? assemblyInfo)
         {
-            if (_mvidMap.TryGetValue(metadataReferenceInfo.ModuleVersionId, out assemblyInfo))
-            {
-                return true;
-            }
+            EnsureCachePopulated(metadataReferenceInfo.FileName);
+            return _mvidMap.TryGetValue(metadataReferenceInfo.ModuleVersionId, out assemblyInfo);
+        }
 
-            if (_nameMap.TryGetValue(metadataReferenceInfo.FileName, out var _))
+        private void EnsureCachePopulated(string fileName)
+        {
+            if (!_nameToLocationsMap.TryGetValue(fileName, out var list))
             {
-                // The file name of this reference has already been searched for and none of them 
-                // had the correct MVID (else the _mvidMap lookup would succeed). No reason to do 
-                // more work here.
-                return false;
+                return;
             }
 
-            var list = new List<AssemblyInfo>();
+            _nameToLocationsMap.Remove(fileName);
 
-            foreach (var directory in _indexDirectories)
+            using var _ = _logger.BeginScope($"Populating {fileName}");
+            var assemblyInfoList = new List<AssemblyInfo>();
```

Comment:
nit: would passing a size hint be helpful here?

Reply:
Usually I only pass a size hint when I know what the size will be. In this case it's hard to know. If you watch the output you'll notice that there are hundreds and hundreds of R2R images that end up getting searched. Using capacity here would end up causing the list to overallocate for a lot of System.* DLLs.